### ISSUE TITLE
Partial dependency between the target and features

### ISSUE THREAD

**Issue by XiaoxiaoWang87** [dmlc/xgboost#486]
Hi,

I'm wondering whether there is a fast way to study the partial dependency between features and the dependent variable. This has been implemented in Scikit learn gradient boosting model and R gbm. 

In R, it is also possible constrain monotonicity to prevent overfitting (for Xgboost which is very effective in preventing overfitting, this is less an issue perhaps). 

In any case I think it would be useful to have a partial dependency functionality in Xgboost as well.

Thanks!

**Comment by jdoepfert** [comment]
I agree, this would indeed by nice!

**Comment by pommedeterresautee** [comment]
I would love implementing it but I read the paper from Friedman (2005) and I have no idea how to compute it from the trees directly. If you can explain me the theory I will do it.

**Comment by XiaoxiaoWang87** [comment]
Thanks for your interests in helping out! Honestly I don't know the exact calculation yet. I may need to look into scikit-learn "sklearn.ensemble.partial_dependenceand" and see how it was implemented there.

**Comment by khotilov** [comment]
There's a generic solution for this in the ICEbox package https://cran.r-project.org/web/packages/ICEbox/
(see http://arxiv.org/abs/1309.6392 for details).

The implementation within mlr framework https://mlr-org.github.io/mlr-tutorial/devel/html/partial_prediction/index.html follows a similar approach and provides somewhat more functionality, albeit constrained to the mlr framework.

If it would make sense, I can add some demo code for how to run xgboost with mlr and to look at partial dependencies.

**Comment by pommedeterresautee** [comment]
In Friedman paper the method described is very light in computation because it leverages the model itself. In these external package there are many computation to perform because it ignores the model. So we need to write it ourselves if we want something which will be used.

**Comment by khotilov** [comment]
@pommedeterresautee could you please point to the description of that Friedman's method? Since the concept of partial dependency is rather model-agnostic and requires re-scoring of a model at many grid points, I'm quite curious to see what tricks Friedman came up with to speed things up.

Note that for a quick approximation for partial dependency, one can look at marginal dependency by calculating predictions only once and then either smoothing or aggregating them in bins WRT features of interest. That's very fast and easy.

And what specifically did you find slow in what those packages offer? 
Here are some tips: 
- one usually wants PD plots for several important features (and not for all the useless features in a model),
- there's no need to use hundreds of thousands of cases to produce such aggregated plots, and xgboost's prediction runs are rather swift with a few 10K's of cases on a multicore machine
- also, plotting huge number of individual predictions could be slow, and there are ways to restrict that number
- when computing pair interaction PD's, grid size should be kept reasonably low
- there seems to be some instability or bug in mlr's derivatives routine and it might take forever, so I would skip that one so far.

**Comment by giuseppec** [comment]
Here is a tutorial how to do this with mlr in general: https://mlr-org.github.io/mlr-tutorial/devel/html/partial_prediction/index.html 
Of course you can also do this using xgboost as learner in mlr:

```r
library(mlr)
library(xgboost)
# define learner
lrn = makeLearner("classif.xgboost", predict.type = "prob", nrounds = 100)

# train model on bh.task
fit = train(lrn, sonar.task)

# get importance
imp = xgb.importance(feature_names = fit$features, model = fit$learner.model)

# make partial dependence plot for most important feature
pd = generatePartialPredictionData(fit, sonar.task, imp$Feature[1])
plotPartialPrediction(pd)
```

![pd](http://i.imgur.com/NHhUIzo.png)

**Comment by aflyax** [comment]
@giuseppec: is there any way to adopt an existing `xgboost` model to be used with `mlr` to get partial dependence plot?

**Comment by giuseppec** [comment]
@aflyax you can try to use the `makeWrappedModel` function from `mlr` and manually create an `mlr` train object (like the `fit` object above). You have to pass your existing `xgboost` model to the `learner.model` function-argument.
I do not guarantee that this will work (not tested) but you can try it out and compare the result with the result after refitting a existing `xgboost` model using `mlr`. If it is the same, it might work.

**Comment by Ishant1589** [comment]
Hi I was using the code shared by giuseppec but was getting the error while using the train function.

Code used:
library(mlr)
library(xgboost)
# define learner

train_label = as.numeric(train1$dependent_var)
train_data = as.matrix(train1[,independent_var])
train_xgmat = xgb.DMatrix(train_data,label = train_label  ,missing = NaN)


lrn = makeLearner("regr.xgboost", predict.type = "response", nrounds = 500)

# train model on bh.task
fit = train(lrn, train_xgmat)

Error:
Error in train(lrn, train_xgmat) : 
  Assertion on 'task' failed: Must have class 'Task', but has class 'xgb.DMatrix'.

Any help would be appreciated

**Comment by giuseppec** [comment]
In your case it should look like:

```
library(mlr)
train_task = makeRegrTask(train1, target ="dependent_var")
fit = train(lrn, train_task)
```
In mlr, you *have* to use Task objects, see also https://mlr-org.github.io/mlr-tutorial/devel/html/task/index.html. You can also do things like `cv = crossval(lrn, train_task, iter = 5)`, etc.

**Comment by Ishant1589** [comment]
@giuseppec  Thanks a lot. The code is running file now.
Also, is it possible to add a watchlist and provide early stopping round using mlr package (similar to what we have in xgb.train function of xgboost package).

Xgboost code below:

```
watchlist <- list(train = train_xgmat, test = test_xgmat)

param = list("objective" = "reg:linear",
             "bst:eta" = 0.05,
             "bst:max_depth" = 6,
             "eval_mteric" = "rmse",
             "silent" = 1,
             "nthread" = 8#,
           #  "num_class" =5
             
)


library(xgboost)
bst <- xgb.train(params = param, data = train_xgmat, nround = 300 , watchlist = watchlist, early_stopping_rounds = 10,verbose=1)


```

**Comment by khotilov** [comment]
@Ishant1589 you can see how mlr wraps xgboost in https://github.com/mlr-org/mlr/blob/master/R/RLearner_classif_xgboost.R

**Comment by zhijiang** [comment]
Will such partial dependency feature be eventually implemented in XGB?  It would be very helpful for us not relying on other packages, as well as the potential gain with the knowledge of the model.  Thanks!

**Comment by bgreenwell** [comment]
I have been working on a partial dependence package, pdp: https://github.com/bgreenwell/pdp. I was able to get it to work with XGBoost, but it is still experimental:

```{r}
# Load required packages
devtools::install_github("bgreenwell/pdp")
library(pdp)
library(xgboost)

# Load the (corrected) Boston housing data 
data(BostonHousing2, package = "mlbench") 
boston <- BostonHousing2[, -c(1, 2, 5)]  # remove unused columns 

# Fit an XGBoost model
set.seed(102) 
boston.xgb <- xgboost(data = data.matrix(subset(boston, select = -cmedv)), 
                                      label = boston$cmedv, 
                                      objective = "reg:linear", 
                                      nrounds = 2000, 
                                      max_depth = 3, 
                                      eta = 0.01)

# PDPs
X <- subset(boston, select = -cmedv)  # remove response!
partial(boston.xgb, pred.var = "lstat", train = X, rug = TRUE)
partial(boston.xgb, pred.var = "rm", train = X, rug = TRUE)
partial(boston.xgb, pred.var = c("lstat", "rm"), chull = TRUE, train = X)
```

**Comment by vinhdizzo** [comment]
Thanks @bgreenwell for the generic package.  Kudos to the two-way partial dependence plot.  Any thoughts on also implementing Friedman's interaction test statistic?

Your package should work well for small data sets, but will probably blow up when the data is large, say half the available memory, due to the number of copies of the X matrix.  I tried something similar for xgboost (calculating partial dependence plots by hand in R), but it blew up with a large data set.  In order for things to scale we might have to do calculations in C to prevent the data copies.

**Comment by bgreenwell** [comment]
Absolutely right, but I never expected this to automatically scale well to LARGE datasets. Hopefully, I (or someone via a PR) can make it more efficient in the future while maintaining the generality. For now, though, the `grid.resolution`, `pred.grid`, and `.parallel` options should provide some relief. Haven't considered Friedman's H-statistic, perhaps I should in the near future!!

**Comment by patr1ckm** [comment]
Here's a relevant snippet from Friedman 2001 on computing a fast partial dependence plot from trees with single variable splits. Let 'z' be the matrix of predictors that we want to plot, and '-z' be the others.

> For regression trees based on single-variable splits, however, the partial dependence of F(x) on a specified target variable subset  is straightforward to evaluate given only the tree, without reference to the data itself. For a specific set of values for the variables z, a weighted traversal of the tree is performed. At the root of the tree, a weight value of 1 is assigned. For each nonterminal node visited, if its split variable is in the target subset z, the appropriate left or right daughter node is visited and the weight is not modified. If the node's split variable is a member of the complement subset (-z), then both daughters are visited and the current weight is multiplied by the fraction of training observations that went left or right, respectively, at that node. 

> Each terminal node visited during the traversal is assigned the current value of the weight. When the tree traversal is complete, the value of \hat{F}(z) is the corresponding weighted average of the F(x) values over those terminal nodes visited during the tree traversal. For a collection of M regression trees, obtained through boosting, the results for the individual trees are simply averaged.

And [here](https://github.com/gbm-developers/gbm/blob/af0660b96fc44191b904555a60102176cb2a9eca/src/gbmentry.cpp) (`gbm_plot`) is the source code in `gbm`. It's very fast.

**Comment by kevin1515** [comment]
when i tried the below:
pd = generatePartialPredictionData(fit, task, imp$Feature[1])
plotPartialPrediction(pd)

I got the error:
Error: could not find function "generatePartialPredictionData"

Could anyone help me? Thanks. 

BTW, i was able to install {lmr} and run belows:
lrn = makeLearner("classif.xgboost", predict.type = "prob", nrounds = 100)
task = makeClassifTask(data = dataTrain, target="Loss")
mdl = train(lrn, task)

**Comment by giuseppec** [comment]
The function `generatePartialPredictionData` has been renamed to `generatePartialDependenceData`, just check the NEWS file from the mlr package.

**Comment by pramitchoudhary** [comment]
There is an open issue in the sklearn repo to have a generalized pdp as well. Overall, I feel that the need to have interpret black box has increase. [here](https://github.com/scikit-learn/scikit-learn/pull/5653). The discussion here and else where has helped me validate this reasoning as well. How do you guys feel about interpretability at global vs individual level as well. 
Recent work on the same: https://arxiv.org/abs/1602.04938.

**Comment by IrinaMax** [comment]
How to extract the best features from the model ?

**Comment by pommedeterresautee** [comment]
@pramitchoudhary you can check this repo if the subject is of interest for you https://github.com/thomasp85/lime
@IrinaMax using feature importance

**Comment by henningsway** [comment]
I currently have the problem, that the output of `plotPartialDependence()` is not a probability for a binary classification-task, even though `predict.type = "prob"` was chosen. 

This is the kind of output I get:
![image](https://user-images.githubusercontent.com/1403304/38033082-6e3ce5a4-329f-11e8-9bc7-1259379391a5.png)

Any ideas? (will try to post a reproducable example soon. :-))

**Comment by bgreenwell** [comment]
Not sure exactly without a reproducible example, but this tends to happen with the [`pdp`](https://github.com/bgreenwell/pdp) package whenever you try to plot a predictor that was never selected as a splitting variable in the ensemble (i.e., its variable importance score is zero).

**Comment by henningsway** [comment]
Thank you @bgreenwell 

Interestingly, the pd object seems to show differences in the level of the probability. I created a short minimal example here: https://stackoverflow.com/questions/49537907/plotting-partial-dependence-plots-in-r-for-binary-target-mlr
